pennylane/tape/tape.py

```diff
@@ -872,6 +872,145 @@ def set_parameters(self, params, trainable_only=True):
+
+            else:
+                shot_vector = device.shots
+                shape = shape = tuple((shot_val,) if shot_val!=1 else tuple() for shot_val in shot_vector)
+
+        return shape, output_domain
+
+    @staticmethod
+    def _multi_same_measurement_shape(mps, device):
+        """
+
+        Note: these shapes and dtypes are specific to tapes and may differ from
+        QNode outputs.
+        """
+
+        shape = tuple()
+
+        output_domain = "real"
+
+        # We know that there's one type of return_type, gather it from the
+        # first one
+        ret_type = mps[0].return_type
+        if ret_type == qml.operation.State:
+            raise ValueError("Multiple state measurements are not supported.")
+
+        if device.shot_vector is None:
+            if ret_type in (qml.operation.Expectation, qml.operation.Variance):
+
+                shape = (len(mps),)
+
+            elif ret_type == qml.operation.Probability:
+
+                wires_num_set = {len(meas.wires) for meas in mps}
+                same_num_wires = len(wires_num_set) == 1
+                if same_num_wires:
+                    # We know that all probability meas processes have the same
+                    # number of wires, gather the length from the first one
+                    dim = 2 ** len(mps[0].wires)
+                    shape = (len(mps), dim)
+
+                else:
+                    # There are varying number of wires that the probability
+                    # measurement processes act on; the output shape is a sum
+                    #
+                    # E.g., for
+                    # return qml.probs(wires=[0]), qml.probs(wires=[1,2])
+                    #
+                    # We'll have a shape of 2 ** 1 + 2 ** 2, where the exponents
+                    # come from the length of the wires specified: 1 == len([0])
+                    # and 2 == len([1, 2])
+                    shape = (sum(2 ** len(m.wires) for m in mps),)
+
+            elif ret_type == qml.operation.Sample:
+
+                output_domain = "integer"
+                shape = (len(mps), device.shots)
+
+        else:
+            if ret_type in (qml.operation.Expectation, qml.operation.Variance):
+                num = sum(shottup.copies for shottup in device.shot_vector)
+                shape = (num, len(mps))
+
+            elif ret_type == qml.operation.Probability:
+                pass
+            elif ret_type == qml.operation.Sample:
+                output_domain = "integer"
+                shape = []
+                for shot_val in device.shot_vector:
+                    for _ in range(shot_val.copies):
+                        shots = shot_val.shots
+                        if shots!=1:
+                            shape.append(tuple([shots,len(mps)]))
+                        else:
+                            shape.append((len(mps),))
+
+        return shape, output_domain
+
+    def get_output_shape_and_domain(self, device):
```

Comment:
@antalszava this strikes me as two methods/properties:

- `tape.output_shape`
- `tape.output_domain`

Reply:
That's a good idea! Separated them like that.